filebeat/input/kafka/input.go

```diff
@@ -0,0 +1,315 @@
+	inputContext input.Context,
+) (input.Input, error) {
+
+	config := defaultConfig()
+	if err := cfg.Unpack(&config); err != nil {
+		return nil, errors.Wrap(err, "reading kafka input config")
+	}
+
+	out, err := connector.ConnectWith(cfg, beat.ClientConfig{
+		Processing: beat.ProcessingConfig{
+			DynamicFields: inputContext.DynamicFields,
+		},
+		ACKEvents: func(events []interface{}) {
+			for _, event := range events {
+				if meta, ok := event.(eventMeta); ok {
+					meta.handler.ack(meta.message)
+				}
+			}
+		},
+		CloseRef:  doneChannelContext(inputContext.Done),
+		WaitClose: config.WaitClose,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	saramaConfig, err := newSaramaConfig(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "initializing Sarama config")
+	}
+
+	input := &kafkaInput{
+		config:       config,
+		saramaConfig: saramaConfig,
+		context:      inputContext,
+		outlet:       out,
+		log:          logp.NewLogger("kafka input").With("hosts", config.Hosts),
+	}
+
+	return input, nil
+}
+
+func (input *kafkaInput) runConsumerGroup() {
+	// Sarama uses standard go contexts to control cancellation, so we need
+	// to wrap our input context channel in that interface.
+	context := doneChannelContext(input.context.Done)
+	handler := &groupHandler{
+		version: input.config.Version,
+		outlet:  input.outlet,
+	}
+
+	// Create a consumer group and make sure it's closed before we return.
+	consumerGroup, err :=
+		sarama.NewConsumerGroup(
+			input.config.Hosts, input.config.GroupID, input.saramaConfig)
+	if err != nil {
+		input.log.Errorw(
+			"Error initializing kafka consumer group", "error", err)
+		return
+	}
+	input.saramaWaitGroup.Add(1)
+	defer func() {
+		consumerGroup.Close()
+		input.saramaWaitGroup.Done()
+	}()
+
+	// Listen asynchronously to any errors during the consume process
+	go func() {
+		for err := range consumerGroup.Errors() {
+			input.log.Errorw("Error reading from kafka", "error", err)
+		}
+	}()
+
+	err = consumerGroup.Consume(context, input.config.Topics, handler)
+	if err != nil {
+		input.log.Errorw("Kafka consume error", "error", err)
+	}
+}
+
+// Run starts the input by scanning for incoming messages and errors.
+func (input *kafkaInput) Run() {
+	input.runOnce.Do(func() {
+		go func() {
+
+			// If the consumer fails to connect, we use exponential backoff with
+			// jitter up to 8 * the initial backoff interval.
+			backoff := backoff.NewEqualJitterBackoff(
+				input.context.Done,
+				input.config.ConnectBackoff,
+				8*input.config.ConnectBackoff)
+			for {
+				// Try to start the consumer group event loop: create a consumer
+				// group client (wbich connects to the kafka cluster) and call
+				// Consume (which starts an asynchronous consumer).
+				input.runConsumerGroup()
+
+				// If runConsumerGroup returns, we wait for the backoff interval.
+				backoff.Wait()
+
+				// Check the Done channel before we try again.
+				select {
+				case <-input.context.Done:
+					return
+				}
```

Comment:
this is missing the `default:` clause, otherwise it blocks forever.
You already created a `context.Context` for the consumer group. If you pass it to the run loop (or generate it again, it's just a type cast), then the for loop becomes:

```
for context.Err() == nil {
   err := input.runConsumerGroup()
   backoff.Wait()
}
```

Where do you call `backoff.Reset()`, so to reset the wait duration to the initial wait time?

Reply:
I didn't call reset in this code -- this was somewhat jumbled because that wait was applying to any fatal error throughout the run loop, so there was no natural place to reset it. I've redistributed the helper code again -- the connection backoff now applies exclusively to the creation / initial connection of the consumer group, and is reset when it succeeds. Errors during the consumption process are subject to sarama's own retry / backoff settings, so when any entire session is lost we just restart the initial connection process right away. Does this make sense?

Reply:
Sounds good.